lib/ex_aws/s3/upload.ex

```diff
@@ -0,0 +1,64 @@
+alias Experimental.GenStage
+
+defmodule ExAws.S3.Upload do
+  @moduledoc """
+  Represents an AWS S3 file download operation
+  """
+
+  @enforce_keys ~w(bucket path src)a
+  defstruct [
+    :src,
+    :bucket,
+    :path,
+    opts: [],
+    service: :s3,
+  ]
+
+  @type t :: %__MODULE__{}
+end
+
+defimpl ExAws.Operation, for: ExAws.S3.Upload do
+
+  alias GenStage.Flow
+
+  def perform(op, config) do
+    upload_id = start_upload(op, config)
+    stream = build_stream(op)
+
+    Flow.new(stages: op.opts[:max_concurrency] || 8, max_demand: 1)
+    |> Flow.from_enumerable(stream)
+    |> Flow.map(&upload_chunk(&1, upload_id, op, config))
+    |> Enum.to_list
+    |> complete_upload(upload_id, op, config)
```

Comment:
Memory usage seems a bit high:
![image](https://cloud.githubusercontent.com/assets/247817/17347097/217a53a8-58dc-11e6-92ca-e3d92bcdb515.png)

How much pre-fetching is it doing?